Re-write ASP.NET Core HTTPS documentation (#6117)
dotnet/dotnet-docker: samples/run-aspnetcore-https-development.md

# Developing ASP.NET Core Applications with Docker over HTTPS

ASP.NET Core uses [HTTPS by
default](https://docs.microsoft.com/aspnet/core/security/enforcing-ssl).
[HTTPS](https://en.wikipedia.org/wiki/HTTPS) relies on
[certificates](https://en.wikipedia.org/wiki/Public_key_certificate) for trust,
identity, and encryption.

This document demonstrates how to develop ASP.NET Core applications with HTTPS
in Docker containers. It's recommended to try the [ASP.NET Core Docker
Sample](README.md) first, which is simpler because the container only exposes
HTTP. This more basic tutorial will help you validate that you have the sample
working correctly, before adding the complication of certificates.

See [Hosting ASP.NET Core images with Docker over HTTPS](https://learn.microsoft.com/en-us/aspnet/core/security/docker-https)
for production scenarios.

The Windows examples below are written for PowerShell. CMD users will need to
change the format of the environment variables in the instructions from
`$env:USERPROFILE` to `%USERPROFILE%`.

> This example requires [Docker
> Desktop](https://www.docker.com/products/docker-desktop/). Make sure you have
> the latest version.

Reviewer: ```suggestion
This example requires [Docker Desktop](https://www.docker.com/products/docker-desktop/). We recommend the latest version.
```

## Getting the sample image

Pull the sample image:

```console
docker pull mcr.microsoft.com/dotnet/samples:aspnetapp
```

Alternatively, you can build the sample image locally:

```console
docker build --pull -t mcr.microsoft.com/dotnet/samples:aspnetapp 'https://github.com/dotnet/dotnet-docker.git#:samples/aspnetapp'
```

## Create and trust a development certificate

> Follow the instructions for creating a [.NET development certificate](https://learn.microsoft.com/dotnet/core/tools/dotnet-dev-certs)
> from [Running pre-built container images with HTTPS](https://learn.microsoft.com/aspnet/core/security/docker-https#running-pre-built-container-images-with-https).

Reviewer: ```suggestion
Follow the instructions for creating a [.NET development certificate](https://learn.microsoft.com/dotnet/core/tools/dotnet-dev-certs) from [Running pre-built container images with HTTPS](https://learn.microsoft.com/aspnet/core/security/docker-https#running-pre-built-container-images-with-https).
```

## Enable HTTPS using environment variables

> See [Hosting ASP.NET Core images with Docker over HTTPS](https://learn.microsoft.com/aspnet/core/security/docker-https).
> If you don't want to use environment variables to store your development
> certificate password, continue reading.

Reviewer: ```suggestion
See [Hosting ASP.NET Core images with Docker over HTTPS](https://learn.microsoft.com/aspnet/core/security/docker-https). See the following section if you don't want to use environment variables to store your development
certificate password.
```

## Using user secrets for certificate password

> If you don't want to use environment variables to store your development
> certificate password, you can use [.NET user secrets](https://learn.microsoft.com/aspnet/core/security/app-secrets?view=aspnetcore-9.0&tabs=linux)
> instead to store the certificate password on your machine.

Reviewer: ```suggestion
Rather than using environment variable to specify the development certificate password,  use [.NET user secrets](https://learn.microsoft.com/aspnet/core/security/app-secrets) to store the password.
```

> This will require modifications to the `aspnetapp` sample, so make sure you've cloned
> this repo or [download the repository as a zip](https://github.com/dotnet/dotnet-docker/archive/main.zip).

Reviewer: ```suggestion
Using User Secrets requires modifications to the `aspnetapp` sample. Clone
this repo or [download the repository as a zip](https://github.com/dotnet/dotnet-docker/archive/main.zip).
```

```console
git clone https://github.com/dotnet/dotnet-docker/
```

> First, initialize user secrets for your app, and set the certificate password.

Reviewer: ```suggestion
Initialize user secrets for your app, and set the certificate password:
```

```console
cd samples/aspnetapp
dotnet user-secrets init -p aspnetapp/aspnetapp.csproj

> dotnet user-secrets -p aspnetapp/aspnetapp.csproj set "Kestrel:Certificates:Default:Password" "<CREDENTIAL_PLACEHOLDER>"

Reviewer: Per SFI, credential instructions can't be copy/paste ready. Add illegal strings so they can't have a working copy/paste. 

```suggestion
dotnet user-secrets -p aspnetapp/aspnetapp.csproj set "Kestrel:Certificates:Default:Password" "<`"CREDENTIAL" "PLACEHOLDER`>"
```

```

> Initializing user-secrets for the first time on a project will modify the
> project file, so we need to re-build the sample image:

Reviewer: ```suggestion
Initializing user-secrets for the first time on a project modifies the project file, which requires re-building the sample image:
```

```pwsh
docker build --pull -t aspnetapp .
```

> In Linux containers, .NET looks under the `~/.microsoft/usersecrets/` directory
> for user secrets data. All you need to do is bind-mount your host machine's
> user secrets directory to the container's filesystem in the correct location,
> similar to what we did with the certificate above. if you are running your
> container as the `root` user, replace `/home/app/` with the `root` user's home
> directory, `/root/`.
> 

Reviewer: ```suggestion
In Linux containers, .NET looks under the `~/.microsoft/usersecrets/` directory for user secrets data. Bind-mount your host machine's user secrets directory to the container's filesystem. This is similar to the preceding certificate instructions. f you are running your container as the `root` user, replace `/home/app/` with the `root` user's home directory, `/root/`.
```

For Linux containers on Windows:

```pwsh
docker run --rm -it `
    -p 8001:8001 `
    -e ASPNETCORE_HTTPS_PORTS=8001 `
    -e ASPNETCORE_ENVIRONMENT=Development `

>     # Use user secrets instead of certificate password environment variable

Reviewer 2: The comments here mess up the multi-line PowerShell. It essentially breaks the line continuations that are used causing an invalid command when pasting this whole thing.

## Enable HTTPS using environment variables

See [Hosting ASP.NET Core images with Docker over HTTPS](https://learn.microsoft.com/aspnet/core/security/docker-https). See the following section if you don't want to use environment variables to store your development
certificate password.

## Using user secrets for certificate password

Rather than using environment variable to specify the development certificate password,  use [.NET user secrets](https://learn.microsoft.com/aspnet/core/security/app-secrets) to store the password.

Initializing user-secrets for the first time on a project modifies the project file, so you will need a local copy of the `aspnetapp` sample. Clone this repo or [download the repository as a zip](https://github.com/dotnet/dotnet-docker/archive/main.zip).

```console
git clone https://github.com/dotnet/dotnet-docker/
```

Initialize user secrets for your app, and set the certificate password:

```console
cd samples/aspnetapp
dotnet user-secrets init -p aspnetapp/aspnetapp.csproj

> dotnet user-secrets -p aspnetapp/aspnetapp.csproj set "Kestrel:Certificates:Default:Password" "<`"CREDENTIAL" "PLACEHOLDER`>"

Reviewer 2: The `<CREDENTIAL_PLACEHOLDER>` change needs to be reverted. It is specifically defined that way as a magic string to appease the secrets scanner.

Author: Can we validate that this new format won't trip the secret scanner? Or find another way to break the command so that it can't be copy-pasted without changes?

Reviewer 2: Don't know. It gets executed automatically on the server side when attempting to push into the AzDO repo.

Author: I was able to push this branch internally without issue: https://dev.azure.com/dnceng/internal/_git/dotnet-dotnet-docker?path=/samples/run-aspnetcore-https-development.md&version=GBdev/loganbussell/pulls/6117&line=51&lineEnd=52&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

Reviewer 2: Ok, that's good. But I don't understand the need for the change. It's not formatted correctly. It leads to this error: `Unrecognized command or argument 'PLACEHOLDER>'`

Author: My interpretation of Rick's suggestion is that the command _shouldn't_ be formatted correctly. It should error if users copy-paste without changing the placeholder text, to prevent them from using `<CREDENTIAL_PLACEHOLDER>` as the password.

Reviewer 2: > My interpretation of Rick's suggestion is that the command _shouldn't_ be formatted correctly. It should error if users copy-paste without changing the placeholder text, to prevent them from using `<CREDENTIAL_PLACEHOLDER>` as the password.

@Rick-Anderson - Is that the case here and if so, is that a common pattern used in our docs? I'm concerned that users will get confused by this syntax and think that things need to be formatted in special way rather than just a simple string.

Reviewer: > > My interpretation of Rick's suggestion is that the command _shouldn't_ be formatted correctly. It should error if users copy-paste without changing the placeholder text, to prevent them from using `<CREDENTIAL_PLACEHOLDER>` as the password.
> 
> @Rick-Anderson - Is that the case here and if so, is that a common pattern used in our docs? I'm concerned that users will get confused by this syntax and think that things need to be formatted in special way rather than just a simple string.

That's the new SFI pattern and mandate. Any markup or code that is copy/pasted must throw an error to prevent copy paste. We can monitor the doc if ppl have problems and add some comments helping them.

Author: I was playing around with other formats. This one works (or rather, doesn't work):

```
> dotnet user-secrets -p aspnetapp/aspnetapp.csproj set "Kestrel:Certificates:Default:Password" $CREDENTIAL_PLACEHOLDER

Missing parameter value for 'value'.
Use the '--help' flag to see info.
```

It also passes AzDO credential check. Since it looks so much better I'm going to update to that.
